### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by YulianaYershova**
Provided unit tests for dao, service, web and http-client layer

### CODE REVIEW

File: dao/src/main/java/com/fitgoal/dao/domain/AuditDto.java [hunk 1]
```diff
@@ -23,6 +19,6 @@
 
     private String event;
 
-    private LocalDateTime date;
+    private Date date;
```
Comment (Ivashchyk) [inline dao/src/main/java/com/fitgoal/dao/domain/AuditDto.java]: Why did you change the date type?

File: dao/src/test/java/com/fitgoal/dao/impl/AuditDaoImplTest.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    @Test
+    public void findAllTest() {
+        // save test data in test DB
+        mongoTestHelper.importDocuments("/audit.json");
+
+        // obtain saved data
+        List<Document> importedDocuments = mongoTestHelper.allDocuments();
+
+        // Convert documents to AuditDto objects
+        List<AuditDto> AuditDtoList = getAuditDtoListFromDocuments(importedDocuments);
```
Comment (Ivashchyk) [inline dao/src/test/java/com/fitgoal/dao/impl/AuditDaoImplTest.java]: Difficult to understand the necessity of all these operations. Could you comment?

File: dao/src/test/java/com/fitgoal/dao/impl/utils/MongoTestHelper.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.List;
+
+@RequiredArgsConstructor
+public class MongoTestHelper {
+
+    @NonNull
+    private final MongoCollection<Document> collection;
+
+    public MongoTestHelper(@NotNull MongoClient mongo, @NotNull String database, @NotNull String collection_name) {
+        this.collection = mongo.getDatabase(database).getCollection(collection_name);
```
Comment (Ivashchyk) [inline dao/src/test/java/com/fitgoal/dao/impl/utils/MongoTestHelper.java]: All constants should be in a camel case

File: http-client/src/main/java/com/fitgoal/BaseClient.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    private final HttpUrl resourceUrl;
+
+    @Inject
+    BaseClient(OkHttpClient okHttpClient, ObjectMapper mapper, HttpUrl resourceUrl) {
+        this.okHttpClient = okHttpClient;
+        this.mapper = mapper;
+        this.resourceUrl = resourceUrl;
+    }
+
+    static ObjectMapper buildDefaultMapper() {
```
Comment (Ivashchyk) [inline http-client/src/main/java/com/fitgoal/BaseClient.java:35]: why is it static? The only reason to have such method is to have an ability to override it in a child class, but it's impossible in case of static. Could you comment?

File: http-client/src/main/java/com/fitgoal/BaseClient.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        }
+    }
+
+    void executeRequest(Request request) {
+        try (Response response = okHttpClient.newCall(request).execute()) {
+            if (!response.isSuccessful()) {
+                throw convertException(response);
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e.getMessage());
```
Comment (Ivashchyk) [inline http-client/src/main/java/com/fitgoal/BaseClient.java]: You will loose a stacktrace, should be simple `new RuntimeException(e)`

File: api/src/main/java/com/fitgoal/api/domain/Recipient.java [hunk 6]
```diff
@@ -1,6 +1,8 @@
 package com.fitgoal.api.domain;
 
+import com.fasterxml.jackson.annotation.JsonProperty;
```
Comment (Ivashchyk) [inline api/src/main/java/com/fitgoal/api/domain/Recipient.java:3]: java.

javax.

org.

com.

other.

**Review by Ivashchyk — CHANGES_REQUESTED** [review]

File: dao/src/main/java/com/fitgoal/dao/domain/AuditDto.java (hunk 1 above)
Comment (YulianaYershova) [reply dao/src/main/java/com/fitgoal/dao/domain/AuditDto.java]: Refactored

File: dao/src/test/java/com/fitgoal/dao/impl/AuditDaoImplTest.java (hunk 2 above)
Comment (YulianaYershova) [reply dao/src/test/java/com/fitgoal/dao/impl/AuditDaoImplTest.java]: Refactored

File: dao/src/test/java/com/fitgoal/dao/impl/utils/MongoTestHelper.java (hunk 3 above)
Comment (YulianaYershova) [reply dao/src/test/java/com/fitgoal/dao/impl/utils/MongoTestHelper.java]: Refactored

File: http-client/src/main/java/com/fitgoal/BaseClient.java (hunk 5 above)
Comment (YulianaYershova) [reply http-client/src/main/java/com/fitgoal/BaseClient.java]: Refactored

File: dao/src/test/java/com/fitgoal/dao/impl/AbstractMongoTestContainersTest.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+import java.util.Collections;
+import java.util.function.Consumer;
+
+public abstract class AbstractMongoTestContainersTest {
+    private static final String MONGO_VERSION = "4.2.1";
+    private static final int MONGO_PORT = 27017;
+    private static final int MONGO_HOST_PORT = 27019;
+
+    private static Consumer<CreateContainerCmd> hostConfig = e -> e.withHostConfig(
```
Comment (shalagatskyi) [inline dao/src/test/java/com/fitgoal/dao/impl/AbstractMongoTestContainersTest.java]: it's a bad practice to use such short and meaningless names and lambda arguments like `e`

File: dao/src/test/java/com/fitgoal/dao/impl/AbstractMongoTestContainersTest.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+import com.github.dockerjava.api.model.ExposedPort;
+import com.github.dockerjava.api.model.PortBinding;
+import com.github.dockerjava.api.model.Ports;
+
+import org.testcontainers.containers.GenericContainer;
+
+import java.util.Collections;
+import java.util.function.Consumer;
+
+public abstract class AbstractMongoTestContainersTest {
```
Comment (shalagatskyi) [inline dao/src/test/java/com/fitgoal/dao/impl/AbstractMongoTestContainersTest.java:20]: great that you extracted the common part in the abstract class!

File: dao/src/test/java/com/fitgoal/dao/impl/utils/DocumentUtils.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+            return documents
+                    .stream()
+                    .map(DocumentUtils::convertDateFromString)
+                    .collect(Collectors.toList());
+        } catch (IOException  ex) {
+            throw new RuntimeException("Can't handle resource: " + resourceName, ex);
+        }
+    }
+
+    private Document convertDateFromString(Document document) {
```
Comment (shalagatskyi) [inline dao/src/test/java/com/fitgoal/dao/impl/utils/DocumentUtils.java]: method name and what it actually does is in confrontation. Moreover it's a bad practice to make hidden mutations like here. Basically you should've had two methods like:
`LocalDateTime convertDate(String date)`
`void fixDate(Document document, LocalDateTime date)`
So basic rule of thumb is that you should follow **Single Responsibility** for methods/classes

File: dao/src/test/java/com/fitgoal/dao/impl/AbstractMongoTestContainersTest.java (hunk 7 above)
Comment (YulianaYershova) [reply dao/src/test/java/com/fitgoal/dao/impl/AbstractMongoTestContainersTest.java]: Refactored

File: dao/src/test/java/com/fitgoal/dao/impl/utils/DocumentUtils.java (hunk 9 above)
Comment (YulianaYershova) [reply dao/src/test/java/com/fitgoal/dao/impl/utils/DocumentUtils.java]: Refactored

File: http-client/src/main/java/com/fitgoal/BaseClient.java (hunk 4 above)
Comment (shalagatskyi) [reply http-client/src/main/java/com/fitgoal/BaseClient.java:35]: agree, there is no need to use the static here. If you wanted to use it for creation of new instance of the BaseClient then the better option would be to create a separate constructor like so:
`
 BaseClient(OkHttpClient okHttpClient, HttpUrl resourceUrl) {
        this.okHttpClient = okHttpClient;
        this.mapper = Jackson.newObjectMapper();
        this.resourceUrl = resourceUrl;
    }
`

File: dao/src/test/java/com/fitgoal/dao/impl/utils/DocumentUtils.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+    private static final SimpleDateFormat SIMPLE_DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+
+    List<Document> parseDocuments(@NonNull String resourceName) {
+        try {
+            String docJson = IOUtils.resourceToString(resourceName, StandardCharsets.UTF_8);
+
+            Object content = Document.parse("{\"json\":" + docJson + "}").get("json");
+            List<Document> documents = (List<Document>) content;
+            return documents
```
Comment (shalagatskyi) [inline dao/src/test/java/com/fitgoal/dao/impl/utils/DocumentUtils.java]: This whole flow looks a bit weird, all you want to do is to add additional field to the existing documents, there is no need to create a new list.
So you can have instead just

`            documents.forEach(fixDate());
            return documents; `
`
    @NotNull
    private static Consumer<Document> fixDate() {
        return document -> document.put(DATE, convertDate(document.getString(DATE)));
    }
`

File: dao/src/test/java/com/fitgoal/dao/impl/utils/DocumentUtils.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+                    .toInstant()
+                    .atZone(ZoneOffset.UTC)
+                    .toLocalDateTime();
+        } catch (ParseException e) {
+            e.printStackTrace();
+        }
+        return localDateTime;
+    }
+
+    private Document fixDate(Document document, LocalDateTime date) {
```
Comment (shalagatskyi) [inline dao/src/test/java/com/fitgoal/dao/impl/utils/DocumentUtils.java]: if you accept my refactoring proposal, remove this method

File: service/src/test/java/com/fitgoal/service/impl/NotificationServiceImplTest.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+@RunWith(MockitoJUnitRunner.class)
+public class NotificationServiceImplTest {
+
+    @Mock
+    private AuditDao auditDao;
+    @Mock
+    private MailSenderImpl mailSenderImpl;
+
+    private NotificationService notificationService;
```
Comment (shalagatskyi) [inline service/src/test/java/com/fitgoal/service/impl/NotificationServiceImplTest.java]: you can use a feature of the MockitoJUnitRunner.class -> just put the @InjectMocks annotation and you can remove the whole @Before method, mockito will create and inject all the @Mock dependencies for you

File: service/src/test/java/com/fitgoal/service/mail/impl/MailSenderImplTest.java [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+@RunWith(MockitoJUnitRunner.class)
+public class MailSenderImplTest {
+
+    @Mock
+    private Mailer mailer;
+    @Mock
+    private SenderConfiguration senderConfiguration;
+
+    private MailSender mailSender;
```
Comment (shalagatskyi) [inline service/src/test/java/com/fitgoal/service/mail/impl/MailSenderImplTest.java]: same here

File: service/src/main/java/com/fitgoal/service/impl/NotificationServiceImpl.java [hunk 14]
```diff
@@ -65,4 +68,15 @@ private void registerEvent(String serviceName, String event) {
                 .date(LocalDateTime.now()).build();
         auditDao.create(auditDto);
     }
+    private URI buildUri(String verificationLink) {
+        try {
+            return new URIBuilder()
+                    .setScheme("http")
```
Comment (shalagatskyi) [inline service/src/main/java/com/fitgoal/service/impl/NotificationServiceImpl.java]: please extract it into a separate config file or reuse MailerConfiguration fields. One of the ways how to do that, is to inject MailerConfiguration in the constructor and store scheme/host/port in the fields.

**Review by shalagatskyi — APPROVED** [review]
looks almost good to me, after fixing my minor comments you can merge this PR

File: dao/src/test/java/com/fitgoal/dao/impl/utils/DocumentUtils.java (hunk 10 above)
Comment (YulianaYershova) [reply dao/src/test/java/com/fitgoal/dao/impl/utils/DocumentUtils.java]: Refactored

File: dao/src/test/java/com/fitgoal/dao/impl/utils/DocumentUtils.java (hunk 11 above)
Comment (YulianaYershova) [reply dao/src/test/java/com/fitgoal/dao/impl/utils/DocumentUtils.java]: removed

File: service/src/test/java/com/fitgoal/service/impl/NotificationServiceImplTest.java (hunk 12 above)
Comment (YulianaYershova) [reply service/src/test/java/com/fitgoal/service/impl/NotificationServiceImplTest.java]: Refactored

File: service/src/test/java/com/fitgoal/service/mail/impl/MailSenderImplTest.java (hunk 13 above)
Comment (YulianaYershova) [reply service/src/test/java/com/fitgoal/service/mail/impl/MailSenderImplTest.java]: Refactored

File: service/src/main/java/com/fitgoal/service/impl/NotificationServiceImpl.java (hunk 14 above)
Comment (YulianaYershova) [reply service/src/main/java/com/fitgoal/service/impl/NotificationServiceImpl.java]: Done
